go/vt/vttablet/tabletserver/query_executor.go

```diff
@@ -716,10 +718,13 @@ func (qre *QueryExecutor) execSelect() (*sqltypes.Result, error) {
 				q.SetErr(err)
 			}
 		} else {
-			qre.logStats.QuerySources |= tabletenv.QuerySourceConsolidator
-			startTime := time.Now()
-			q.Wait()
-			qre.tsv.stats.WaitTimings.Record("Consolidations", startTime)
+			waiterCap := qre.tsv.config.ConsolidatorQueryWaiterCap
+			if waiterCap == 0 || c.WaiterCountOfTotal() <= waiterCap {
+				qre.logStats.QuerySources |= tabletenv.QuerySourceConsolidator
+				startTime := time.Now()
+				q.Wait()
+				qre.tsv.stats.WaitTimings.Record("Consolidations", startTime)
+			}
```

Comment:
Adding the waiter in `Record` is too late. As the wait is already complete in `q.Wait()`.
Post that `q.Result()` returns the stored result.

`c.WaiterCountOfTotal() <= waiterCap` - This check will allow concurrent threads to go through without respecting the limit.

Reply:
Hi @harshit-gangal Could you please let me know if I understand correctly regarding your concern.

>c.WaiterCountOfTotal() <= waiterCap - This check will allow concurrent threads to go through without respecting the limit.

- Same query won't run concurrently, because [q](https://github.com/vitessio/vitess/blob/e39321db0a725288cd0fd0551c5c5cbc46096408/go/vt/vttablet/tabletserver/query_executor.go#L704-L706) is a same ***pendingResult** object

- Different queries can run concurrently, however, my solution use **totalWaiterCount** and only update those queries that are actually going to Wait() for ***pendingResult**. For example, if there are 2 different queries, and:
-- select a from A (1 running, 100 waiting)
-- select b from B (1 running, 100 waiting)
Then **totalWaiterCount** is 100+100=200, I added testCase on this.

>c.WaiterCountOfTotal() <= waiterCap - This check will allow concurrent threads to go through without respecting the limit.

WaiterCountOfTotal() is a concurrent safe reader, so that should be respected? 
```
func (co *consolidator) WaiterCountOfTotal() int64 {
	return atomic.LoadInt64(co.totalWaiterCount)
}
```

>Adding the waiter in Record is too late. As the wait is already complete in q.Wait().
Post that q.Result() returns the stored result.

`Record` is already used to [increase the count of query](https://github.com/vitessio/vitess/blob/252d5327e94874018835c3d0ca0d95f494d3d8c4/go/sync2/consolidator.go#L162), for queries passed `c.WaiterCountOfTotal() <= waiterCap` check, seems a good space to increase **totalWaiterCount** as well.